rag/chains.py

```diff
@@ -0,0 +1,34 @@
+from rag.vector_databases import chatbot_retriever
+from rag.prompt_templates import chatbot_template, chatbot_with_history_template
+from rag.utils.llm import openai_llm
+from rag.utils.text_processing import format_source_docs
+
+from langchain_core.output_parsers import StrOutputParser
+from langchain_core.runnables import RunnablePassthrough, RunnableParallel
+
+from rag.utils.history import get_session_history
+from langchain_core.runnables.history import RunnableWithMessageHistory
+
+rag_chain = chatbot_template | openai_llm | StrOutputParser()
+
+rag_with_source_base = RunnableParallel(
+    {
+        "context": chatbot_retriever | format_source_docs,
+        "query": RunnablePassthrough(),
+    }
+)
+
+rag_chain_with_source = rag_with_source_base.assign(
+    answer=rag_chain
+)  # Chain assignments aren't dynamically configurable yet
+
+rag_with_history_chain = chatbot_with_history_template | openai_llm | StrOutputParser()
+
+chatbot_with_history_chain = rag_with_source_base.assign(
```

Comment:
is my understanding correct that the RAG chain with History will add the history to the prompt (i.e. we have the prompt + messaging history)?

Reply:
there is a basic implementation of RAG with in-memory message history. it would better to use a Redis implementation but this is a nice to have/ not in the current scope.